// rna-visualizer/dist/rna/rectangle.js
import { identity, } from './data-structures';
var Rectangle = /** @class */ (function () {
    function Rectangle(coor, width, height, classes) {
        this.visible = true;
        this.transform = identity;
        this.classes = classes;
        this.coor = coor;
        this.width = width;
        this.height = height;
    }
    Rectangle.prototype.setTransform = function (transform) {
        this.transform = transform;
        return this;
    };
    Rectangle.prototype.getTransformedX = function () {
        return this.transform.applyX(this.coor.x);
    };
    Rectangle.prototype.getX = function () {
        return this.coor.x;
    };
    Rectangle.prototype.setX = function (x) {
        this.coor.x = x;
        return this;
    };
    Rectangle.prototype.getTransformedY = function () {
        return this.transform.applyY(this.coor.y);
    };
    Rectangle.prototype.getY = function () {
        return this.coor.y;
    };
    Rectangle.prototype.setY = function (y) {
        this.coor.y = y;
        return this;
    };
    Rectangle.prototype.setCoor = function (coor) {
        this.coor = coor;
        return this;
    };
    Rectangle.prototype.getCoor = function () {
        return this.coor.copy();
    };
    Rectangle.prototype.setVisible = function (visible) {
        this.visible = visible;
        return this;
    };
    Rectangle.prototype.isVisible = function () {
        return this.visible;
    };
    Rectangle.prototype.getWidth = function () {
        return this.width;
    };
    Rectangle.prototype.getHeight = function () {
        return this.height;
    };
    Rectangle.prototype.getClasses = function () {
        return this.classes;
    };
    Rectangle.prototype.translate = function (shift) {
        this.coor.add(shift);
        return this;
    };
    return Rectangle;
}());
export { Rectangle };
