// rna-visualizer/dist/rna/vector.js
var Vector2 = /** @class */ (function () {
    function Vector2(x, y) {
        this.x = x;
        this.y = y;
    }
    Vector2.prototype.add = function (vector) {
        this.x += vector.x;
        this.y += vector.y;
        return this;
    };
    Vector2.prototype.subtract = function (vector) {
        this.x -= vector.x;
        this.y -= vector.y;
        return this;
    };
    Vector2.prototype.multiply = function (k) {
        this.x *= k;
        this.y *= k;
        return this;
    };
    Vector2.prototype.size = function () {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
    };
    Vector2.prototype.copy = function () {
        return new Vector2(this.x, this.y);
    };
    Vector2.sum = function (vector1, vector2) {
        return new Vector2(vector1.x, vector1.y).add(vector2);
    };
    Vector2.subtraction = function (vector1, vector2) {
        return new Vector2(vector1.x, vector1.y).subtract(vector2);
    };
    Vector2.distance = function (vector1, vector2) {
        return this.subtraction(vector1, vector2).size();
    };
    return Vector2;
}());
export { Vector2 };
