// rna-visualizer/dist/components/label.js
/**
* Label class.
*/
export class Label {
    /**
    * @param residue - Residue to which the label is attached.
    * @param line - Line component of the label.
    * @param text - Text component of the label.
    */
    constructor(residue, line, text) {
        this.visible = true;
        this.residue = residue;
        this.line = line;
        this.text = text;
    }
    /**
    * @param transform - Transformation to apply to the label.
    */
    setTransform(transform) {
        this.line.setTransform(transform);
        this.text.setTransform(transform);
        return this;
    }
    /**
    * @param visible - Whether the label should be visible.
    */
    setVisible(visible) {
        this.line.setVisible(visible);
        this.text.setVisible(visible);
        this.visible = visible;
        return this;
    }
    /**
    * @returns Whether the label is visible.
    */
    isVisible() {
        return this.visible && this.residue.isVisible();
    }
    /**
    * @param shift - Vector to translate the label by.
    * @returns The label.
    */
    translate(shift) {
        this.line.translate(shift);
        this.text.translate(shift);
        return this;
    }
}
