// rna-visualizer/dist/components/line.js
import { identity, } from '../components';
/**
 * A class representing a line segment between two points.
 */
export class Line {
    /**
     * Create a new Line object.
     * @param coor1 - The first coordinate of the line segment.
     * @param coor2 - The second coordinate of the line segment.
     * @param classes - An array of classes for styling the line segment.
     */
    constructor(coor1, coor2, classes) {
        this.visible = true;
        this.transform = identity;
        this.coor1 = coor1;
        this.coor2 = coor2;
        this.classes = classes;
    }
    /**
     * Set the transformation to apply to this line segment.
     * @param transform - The transformation to apply.
     * @returns This line segment, for chaining.
     */
    setTransform(transform) {
        this.transform = transform;
        return this;
    }
    /**
     * Get the transformed X coordinate of the first point of this line segment.
     * @returns The transformed X coordinate of the first point.
     */
    getTransformedX1() {
        return this.transform.applyX(this.coor1.x);
    }
    /**
     * Get the transformed Y coordinate of the first point of this line segment.
     * @returns The transformed Y coordinate of the first point.
     */
    getTransformedY1() {
        return this.transform.applyY(this.coor1.y);
    }
    /**
     * Get the transformed X coordinate of the second point of this line segment.
     * @returns The transformed X coordinate of the second point.
     */
    getTransformedX2() {
        return this.transform.applyX(this.coor2.x);
    }
    /**
     * Get the transformed Y coordinate of the second point of this line segment.
     * @returns The transformed Y coordinate of the second point.
     */
    getTransformedY2() {
        return this.transform.applyY(this.coor2.y);
    }
    /**
     * Set the X coordinate of the first point of this line segment.
     * @param x1 - The new X coordinate of the first point.
     * @returns This line segment, for chaining.
     */
    setX1(x1) {
        this.coor1.x = x1;
        return this;
    }
    /**
     * Get the X coordinate of the first point of this line segment.
     * @returns The X coordinate of the first point.
     */
    getX1() {
        return this.coor1.x;
    }
    /**
     * Set the Y coordinate of the first point of this line segment.
     * @param y1 - The new Y coordinate of the first point.
     * @returns This line segment, for chaining.
     */
    setY1(y1) {
        this.coor1.y = y1;
        return this;
    }
    /**
     * Get the Y coordinate of the first point of this line segment.
     * @returns The Y coordinate of the first point.
     */
    getY1() {
        return this.coor1.y;
    }
    /**
     * Set the X coordinate of the second point of this line segment.
     * @param x2 - The new X coordinate of the second point.
     * @returns This line segment, for chaining.
     */
    setX2(x2) {
        this.coor2.x = x2;
        return this;
    }
    /**
     * Gets the x-coordinate of the second point of the line.
     */
    getX2() {
        return this.coor2.x;
    }
    /**
     * Sets the Y coordinate of the second point of the line.
     * @param y2 - The new Y coordinate.
     */
    setY2(y2) {
        this.coor2.y = y2;
        return this;
    }
    /**
     * Gets the Y coordinate of the second point of the line.
     */
    getY2() {
        return this.coor2.y;
    }
    /**
     * Sets the first point of the line to the given Vector2.
     * @param coor - The new coordinates.
     */
    setCoor1(coor) {
        this.coor1 = coor;
        return this;
    }
    /**
     * Gets a copy of the first point of the line as a Vector2.
     */
    getCoor1() {
        return this.coor1.copy();
    }
    /**
     * Sets the second point of the line to the given Vector2.
     * @param coor - The new coordinates.
     */
    setCoor2(coor) {
        this.coor2 = coor;
        return this;
    }
    /**
     * Gets a copy of the second point of the line as a Vector2.
     */
    getCoor2() {
        return this.coor2.copy();
    }
    /**
     * Sets the visibility of the line.
     * @param visible - True if the line should be visible, false otherwise.
     */
    setVisible(visible) {
        this.visible = visible;
        return this;
    }
    /**
     * Gets the visibility of the line.
     */
    isVisible() {
        return this.visible;
    }
    /**
     * Gets the classes applied to the line.
     */
    getClasses() {
        return this.classes;
    }
    /**
     * Translates the line by the given Vector2.
     * @param shift - The Vector2 to translate by.
     */
    translate(shift) {
        this.coor1.add(shift);
        this.coor2.add(shift);
        return this;
    }
}
