// rna-visualizer/dist/components/mappingLine.js
/**
 * Represents a line between two residues that are mapped to each other.
 */
export class MappingLine {
    /**
     * Constructs a MappingLine object
     * @param residue1 - the first residue object
     * @param residue2 - the second residue object
     * @param classes - an array of strings representing the classes of the MappingLine object for styling the line.
     */
    constructor(residue1, residue2, classes) {
        this.visible = false;
        this.residue1 = residue1;
        this.residue2 = residue2;
        this.classes = classes;
    }
    /**
     * Gets the transformed X coordinate of the first residue
     * @returns a number representing the transformed X coordinate of the first residue
     */
    getTransformedX1() {
        return this.residue1.getTransformedX();
    }
    /**
     * Gets the transformed Y coordinate of the first residue
     * @returns a number representing the transformed Y coordinate of the first residue
     */
    getTransformedY1() {
        return this.residue1.getTransformedY();
    }
    /**
     * Gets the transformed X coordinate of the second residue
     * @returns a number representing the transformed X coordinate of the second residue
     */
    getTransformedX2() {
        return this.residue2.getTransformedX();
    }
    /**
     * Gets the transformed Y coordinate of the second residue
     * @returns a number representing the transformed Y coordinate of the second residue
     */
    getTransformedY2() {
        return this.residue2.getTransformedY();
    }
    /**
     * Gets the X coordinate of the first residue
     * @returns a number representing the X coordinate of the first residue
     */
    getX1() {
        return this.residue1.getX();
    }
    /**
     * Gets the Y coordinate of the first residue
     * @returns a number representing the Y coordinate of the first residue
     */
    getY1() {
        return this.residue1.getY();
    }
    /**
     * Gets the X coordinate of the second residue
     * @returns a number representing the X coordinate of the second residue
     */
    getX2() {
        return this.residue2.getX();
    }
    /**
     * Gets the Y coordinate of the second residue
     * @returns a number representing the Y coordinate of the second residue
     */
    getY2() {
        return this.residue2.getY();
    }
    /**
     * Sets the visibility of the MappingLine object
     * @param visible - a boolean value representing the visibility of the MappingLine object
     */
    setVisible(visible) {
        this.visible = visible;
    }
    /**
     * Gets the visibility of the MappingLine object
     * @returns a boolean value representing the visibility of the MappingLine object
     */
    isVisible() {
        return this.visible;
    }
    /**
     * Gets an array of strings representing the classes of the MappingLine object
     * @returns an array of strings representing the classes of the MappingLine object
     */
    getClasses() {
        return this.classes;
    }
    /**
    * Creates MappingLine objects for each residue pair that can be mapped between template and derived container.
    * @param template - The DataContainer representing the template structure.
    * @param container - The DataContainer representing the derived structure.
    * @param classes - (Optional) An array of strings to set as the classes property for each MappingLine object.
    * @returns An array of MappingLine objects representing the residue mappings between template and derived container.
    */
    static createMappingLines(template, container, classes = ['mapping-line']) {
        const mappingLines = [];
        container.getMappableResidues().forEach(res => {
            const tempRes = template.residues[res.templateIndex];
            const mp = new MappingLine(tempRes, res, classes);
            mappingLines.push(mp);
        });
        return mappingLines;
    }
}
