// rna-visualizer/dist/components/residue.js
import { Vector2, Text, Circle, } from '../components';
import { Styles } from '../styles';
/**
 * Class representing a Residue
 */
export class Residue {
    /**
     * Create a Residue.
     * @param name - The name of the residue.
     * @param index - The index of the residue.
     * @param templateName - The name of the template residue.
     * @param templateIndex - The index of the template residue.
     * @param circle - The Circle object representing the text background.
     * @param text - The Text object representing the residue name.
     */
    constructor(name, index, templateName, templateIndex, circle, text) {
        this.visible = true;
        this.name = name;
        this.index = index;
        this.templateIndex = templateIndex;
        this.templateName = templateName;
        this.circle = circle;
        this.text = text;
    }
    /**
     * Creates a Residue from IDataResidue object.
     * @param res - The IDataResidue object representing a residue.
     * @param styles - The Styles object representing the styles of the residue.
     * @returns A Residue object created from the given IDataResidue object.
     */
    static fromDataResidue(res, styles) {
        const textCoor = new Vector2(res.x, res.y);
        const classes = Object.assign([], res.classes);
        classes.push(Styles.TRANSFORMED_CLASS);
        const text = new Text(textCoor, res.residueName, classes);
        const circleCoor = new Vector2(res.x, res.y);
        const radius = Number(styles.getProperty(res.classes, 'font-size').slice(0, -2)) * 0.75;
        const circle = new Circle(circleCoor, radius);
        return new Residue(res.residueName, res.residueIndex, res.templateResidueName, res.templateResidueIndex, circle, text);
    }
    /**
     * Sets the transformation for the Residue.
     * @param transform - The Transformation object representing the transformation to set.
     * @returns The Residue object.
     */
    setTransform(transform) {
        this.text.setTransform(transform);
        this.circle.setTransform(transform);
        return this;
    }
    /**
     * Gets the transformed X coordinate of the Residue.
     * @returns The transformed X coordinate of the Residue.
     */
    getTransformedX() {
        return this.circle.getTransformedX();
    }
    /**
     * Gets the X coordinate of the Residue.
     * @returns The X coordinate of the Residue.
     */
    getX() {
        return this.circle.getX();
    }
    /**
     * Sets the X coordinate of the Residue.
     * @param x - The X coordinate to set.
     * @returns The Residue object.
     */
    setX(x) {
        this.circle.setX(x);
        this.text.setX(x);
        return this;
    }
    /**
     * Gets the transformed Y coordinate of the Residue.
     * @returns The transformed Y coordinate of the Residue.
     */
    getTransformedY() {
        return this.circle.getTransformedY();
    }
    /**
     * Gets the Y coordinate of the Residue.
     * @returns The Y coordinate of the Residue.
     */
    getY() {
        return this.circle.getY();
    }
    /**
     * Sets the Y coordinate of the Residue.
     * @param y - The Y coordinate to set.
     * @returns The Residue object.
     */
    setY(y) {
        this.circle.setY(y);
        this.text.setY(y);
        return this;
    }
    /**
     * Sets the coordinate of the Residue.
     * @param coor - The Vector2 object representing the coordinate to set.
     * @returns The Residue object.
     */
    setCoor(coor) {
        this.circle.setCoor(coor);
        this.text.setCoor(coor);
        return this;
    }
    /**
     * Gets the coordinate of the Residue.
     * @returns The Vector2 object representing the coordinate of the Residue.
     */
    getCoor() {
        return this.circle.getCoor();
    }
    /**
     * Gets the transformed coordinate of the Residue.
     * @returns The Vector2 object representing the transformed coordinate of the Residue.
     */
    getTransformedCoor() {
        return this.circle.getTransformedCoor();
    }
    /**
     * Sets the visibility of the Residue.
     * @param visible - The boolean value representing the visibility to set.
     * @returns The Residue object.
     */
    setVisible(visible) {
        this.circle.setVisible(visible);
        this.text.setVisible(visible);
        this.visible = visible;
        return this;
    }
    /**
     * Gets the visibility of the Residue.
     * @returns The boolean value representing the visibility of the Residue.
     */
    isVisible() {
        return this.visible;
    }
    /**
     * Gets the classes of the Residue text.
     * @returns The array of strings representing the classes of the Residue text.
     */
    getClasses() {
        return this.text.getClasses();
    }
    /**
    * Translates the residue's circle and text by the given shift vector.
    *
    * @param shift - The vector to shift the residue's circle and text by.
    * @returns This Residue instance.
    */
    translate(shift) {
        this.circle.translate(shift);
        this.text.translate(shift);
        return this;
    }
}
