// rna-visualizer/dist/components/vector.js
/**
 * Two-dimensional vector.
 */
export class Vector2 {
    /**
     * Creates a new Vector2 instance.
     * @param x - The x component of the vector.
     * @param y - The y component of the vector.
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
    /**
     * Adds the given vector to this vector.
     * @param vector - The vector to add.
     * @returns The sum of this vector and the given vector.
     */
    add(vector) {
        this.x += vector.x;
        this.y += vector.y;
        return this;
    }
    /**
     * Subtracts the given vector from this vector.
     * @param vector - The vector to subtract.
     * @returns The difference between this vector and the given vector.
     */
    subtract(vector) {
        this.x -= vector.x;
        this.y -= vector.y;
        return this;
    }
    /**
     * Multiplies this vector by the given scalar.
     * @param k - The scalar to multiply by.
     * @returns This vector multiplied by the given scalar.
     */
    multiply(k) {
        this.x *= k;
        this.y *= k;
        return this;
    }
    /**
     * Returns the size (magnitude) of this vector.
     * @returns The size of this vector.
     */
    size() {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
    }
    /**
     * Returns a copy of this vector.
     * @returns A copy of this vector.
     */
    copy() {
        return new Vector2(this.x, this.y);
    }
    /**
     * Returns the sum of the given vectors.
     * @param vector1 - The first vector.
     * @param vector2 - The second vector.
     * @returns The sum of the given vectors.
     */
    static sum(vector1, vector2) {
        return new Vector2(vector1.x, vector1.y).add(vector2);
    }
    /**
     * Returns the difference between the given vectors.
     * @param vector1 - The first vector.
     * @param vector2 - The second vector.
     * @returns The difference between the given vectors.
     */
    static subtraction(vector1, vector2) {
        return new Vector2(vector1.x, vector1.y).subtract(vector2);
    }
    /**
     * Returns the distance between the given vectors.
     * @param vector1 - The first vector.
     * @param vector2 - The second vector.
     * @returns The distance between the given vectors.
     */
    static distance(vector1, vector2) {
        return this.subtraction(vector1, vector2).size();
    }
}
/**
 * The zero vector.
 */
Vector2.zero = new Vector2(0, 0);
