// rna-visualizer/dist/mappingLines.js
var MappingLine = /** @class */ (function () {
    function MappingLine(residue1, residue2) {
        this.classes = ['', ''];
        this.visible = true;
        this.residue1 = residue1;
        this.residue2 = residue2;
    }
    MappingLine.prototype.getTransformedX1 = function () {
        return this.residue1.getTransformedX();
    };
    MappingLine.prototype.getTransformedY1 = function () {
        return this.residue1.getTransformedY();
    };
    MappingLine.prototype.getTransformedX2 = function () {
        return this.residue2.getTransformedX();
    };
    MappingLine.prototype.getTransformedY2 = function () {
        return this.residue2.getTransformedY();
    };
    MappingLine.prototype.getX1 = function () {
        return this.residue1.getX();
    };
    MappingLine.prototype.getY1 = function () {
        return this.residue1.getY();
    };
    MappingLine.prototype.getX2 = function () {
        return this.residue2.getX();
    };
    MappingLine.prototype.getY2 = function () {
        return this.residue2.getY();
    };
    MappingLine.prototype.setVisible = function (visible) {
        this.visible = visible;
    };
    MappingLine.prototype.isVisible = function () {
        return this.visible;
    };
    MappingLine.prototype.getClasses = function () {
        return this.classes;
    };
    MappingLine.CreateMappingLines = function (template, container) {
        var mappingLines = [];
        container.getMappableResidues().forEach(function (res) {
            var tempRes = template.residues[res.templateIndex];
            var mp = new MappingLine(tempRes, res);
            mappingLines.push(mp);
        });
        return mappingLines;
    };
    return MappingLine;
}());
export { MappingLine };
